File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        if(s.length()<p.length()){
            return new ArrayList<>();
        }
        HashMap<Character,Integer> mp1=new HashMap<>();
        for(char ch:p.toCharArray()){
            mp1.put(ch,mp1.getOrDefault(ch,0)+1);
        }
        int left=0;
        HashMap<Character,Integer> mp2=new HashMap<>();
        List<Integer> list=new ArrayList<>();
        for(int right=0;right<s.length();right++){
            char ch=s.charAt(right);
            mp2.put(ch,mp2.getOrDefault(ch,0)+1);
            while(right-left+1>p.length()){
                mp2.put(s.charAt(left),mp2.get(s.charAt(left))-1);
                if(mp2.get(s.charAt(left))==0){
                    mp2.remove(s.charAt(left));
                }
                left++;
            }
            if(mp1.equals(mp2)){
                list.add(left);
            }
        }
        return list;
    }
}

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if(s.length()<p.length()){
            return vector<int>{};
        }
        map<char,int> mp1;
        for(char ch:p){
            mp1[ch]++;
        }
        int left=0;
        map<char,int> mp2;
        vector<int> list;
        for(int right=0;right<s.length();right++){
            mp2[s[right]]++;
            while(right-left+1>p.length()){
                mp2[s[left]]--;
                if(mp2[s[left]]==0){
                    mp2.erase(s[left]);
                }
                left++;
            }
            if(mp1==mp2){
                list.push_back(left);
            }
        }
        return list;
    }
};